apps/emqx/src/emqx_persistent_session_ds_state.erl

```diff
@@ -478,94 +592,221 @@ pmap_del(
 pmap_format(#pmap{cache = Cache}) ->
     Cache.
 
 %% Functions dealing with set tables:
 
-kv_persist(Tab, SessionId, Val0) ->
-    Val = encoder(encode, Tab, Val0),
-    mnesia:write(Tab, #kv{k = SessionId, v = Val}, write).
+session_restore(SessionId) ->
+    Empty = maps:from_keys(
+        [
+            ?metadata_domain,
+            ?subscription_domain,
+            ?stream_domain,
+            ?rank_domain,
+            ?seqno_domain
+        ],
+        []
+    ),
+    Data = maps:groups_from_list(
+        fun(#{domain := Domain}) ->
+            Domain
+        end,
+        read_iterate(SessionId, ['#'])
+    ),
+    maps:merge(Empty, Data).
 
-kv_restore(Tab, SessionId) ->
-    [encoder(decode, Tab, V) || #kv{v = V} <- mnesia:read(Tab, SessionId)].
+%%
 
-%% Functions dealing with bags:
+to_domain_msg(Domain, SessionId, Key, Val) ->
+    #message{
+        %% unused; empty binary to satisfy dialyzer
+        id = <<>>,
+        timestamp = 0,
+        from = SessionId,
+        topic = domain_topic_encode(Domain, SessionId, Key),
+        payload = val_encode(Domain, Val)
+    }.
 
-%% @doc Create a mnesia table for the PMAP:
--spec create_kv_pmap_table(atom()) -> ok.
-create_kv_pmap_table(Table) ->
-    mria:create_table(Table, [
-        {type, ordered_set},
-        {rlog_shard, ?DS_MRIA_SHARD},
-        {storage, rocksdb_copies},
-        {record_name, kv},
-        {attributes, record_info(fields, kv)}
+from_domain_msg(#message{topic = Topic, payload = Bin}) ->
+    #{
+        domain := Domain,
+        session_id := _SessionId,
+        key := _Key
+    } = Data = domain_topic_decode(Topic),
+    Data#{val => val_decode(Domain, Bin)}.
+
+-spec domain_topic_encode
+    (
+        ?metadata_domain,
+        emqx_persistent_session_ds:id(),
+        _Unused :: any()
+    ) ->
+        binary();
+    (
+        ?subscription_domain,
+        emqx_persistent_session_ds:id(),
+        term()
+    ) ->
+        binary();
+    (
+        ?stream_domain,
+        emqx_persistent_session_ds:id(),
+        {term(), emqx_ds:stream()}
+    ) ->
+        binary();
+    (
+        ?rank_domain,
+        emqx_persistent_session_ds:id(),
+        rank_key()
+    ) ->
+        binary();
+    (
+        ?seqno_domain,
+        emqx_persistent_session_ds:id(),
+        seqno_type()
+    ) ->
+        binary().
+domain_topic_encode(Domain, SessionId, Key) when
+    Domain =:= ?metadata_domain;
+    Domain =:= ?subscription_domain;
+    Domain =:= ?stream_domain;
+    Domain =:= ?rank_domain;
+    Domain =:= ?seqno_domain
+->
+    %% This introduces a redundant extra topic level for the metadata domain, but it also
+    %% makes the topic structure uniform in relation to the other domains.
+    BinKey = key_encode(Domain, Key),
+    to_topic(Domain, SessionId, BinKey).
+
+to_topic(Domain, SessionId0, BinKey) when is_binary(BinKey) ->
+    SessionId = emqx_http_lib:uri_encode(SessionId0),
+    emqx_topic:join([
+        <<"session">>,
```

Comment:
I don't see the reason to have a common prefix for the topics.

Reply:
I think it's a good idea to have namespaces, in case we might want to add other data to this database, such as some kind of metadata not tied to a particular clientid.